mono/mini/mini-wasm-debugger.c

```diff
@@ -738,35 +758,69 @@ describe_object_properties (guint64 objectId, gboolean isAsyncLocalThis)
 			continue;
 		if (mono_field_is_deleted (f))
 			continue;
-		mono_wasm_add_properties_var(f->name);
-		gpointer field_value = (guint8*)obj + f->offset;
+		if (g_str_has_suffix (f->name, "k__BackingField"))
+			continue;
+
+		mono_wasm_add_properties_var (f->name, f->offset);
+
+		gpointer field_addr;
+		if (is_valuetype)
+			field_addr = mono_vtype_get_field_addr (obj, f);
+		else
+			field_addr = (guint8*)obj + f->offset;
 		
-		describe_value(f->type, field_value);
+		describe_value (f->type, field_addr, is_valuetype | expandValueType);
 	}
 
 	iter = NULL;
-	while ((p = mono_class_get_properties (obj->vtable->klass, &iter))) {
+	pnum = 0;
+	while ((p = mono_class_get_properties (klass, &iter))) {
 		DEBUG_PRINTF (2, "mono_class_get_properties - %s - %s\n", p->name, p->get->name);
 		if (p->get->name) { //if get doesn't have name means that doesn't have a getter implemented and we don't want to show value, like VS debug
-			if (isAsyncLocalThis &&  (p->name[0] != '<' || (p->name[0] == '<' &&  p->name[1] == '>'))) {
+			if (isAsyncLocalThis && (p->name[0] != '<' || (p->name[0] == '<' &&  p->name[1] == '>')))
 				continue;
-			}
-			mono_wasm_add_properties_var(p->name); 
+
+			mono_wasm_add_properties_var (p->name, pnum);
 			sig = mono_method_signature_internal (p->get);
+			if (is_valuetype && mono_class_from_mono_type_internal (sig->ret) == klass) {
+				// Property of the same valuetype, avoid endlessly recursion!
+				mono_wasm_add_string_var (mono_class_full_name (klass));
+				continue;
+			}
+
 			res = mono_runtime_try_invoke (p->get, obj, NULL, &exc, error);
 			if (!is_ok (error) && exc == NULL)
```

Comment:
This code looks suspect, invoking property accessors could hit breakpoints, hang on locks, etc. so it should not be done here.

Reply:
I was thinking that too, but VS seems to invoke the getters when enumerating the members of the containing object. I tried that by adding a CWL to the getter, and that gets logged to the console. If I add a `Thread.Sleep (10000)`, then the debugger takes time to expand the value.

Reply:
I can restrict to automatic properties for now. And devtools-protocol supports explict getters, so we can maybe use that.

Reply:
@radical good idea, there are some debugger specific attributes we should also honor but autos and fields sound good here.

Reply:
The IDEs have an option to enable/disable this. Also when its enabled, the IDE invokes the getters one by one with a timeout. Also, the ide can disassemble the getters and run the simple ones on the IDE side, i.e. if the getter is an ldfld, then it can query the field directly.

Reply:
Ah, Thanks for the explanation!